lib/cchain/provider/cache.go

```diff
@@ -0,0 +1,86 @@
+package provider
+
+import (
+	"sync"
+
+	"github.com/omni-network/omni/lib/xchain"
+)
+
+// chainVersionCache contains approved attestations of a chain version with all heights above the watermark.
+type chainVersionCache struct {
+	cache     map[uint64]xchain.Attestation
```

Comment:
I even thought about it, but decided against to avoid looking up the insertion point, plus handling of slice mutation (I remember slices have a lot of gotchas)